### DISCUSSION TITLE
Support is_true / is_false methods for boolean expressions

### DISCUSSION THREAD

**Discussion by belovaf** [diesel-rs/diesel#4090] [Ideas]
Even crates.io has a lot of `booleanProp.eq(true)` expressions, for example:
[github_secret_scanning.rs#L62](https://github.com/rust-lang/crates.io/blob/0c8c7c1d74d516349851a85320371acac765141c/src/tests/github_secret_scanning.rs#L62)
This leads to an additional bind parameter and slows down a query.
It would be very convenient (and better performance) to have functions `is_true` / `is_false` to be rendered as `= true` / `= false`.

**Comment by weiznich** [comment]
I would like to see hard numbers for the claim that additional bind parameters slow down the query. Otherwise this does not seem to be justified in any way, given that it would require us to start dealing to convert rust values to SQL string representations with is something that we really want to avoid.

**Reply by belovaf** [reply]
Postgres (other databases have similar concepts) have 2 different plan types: [generic and custom](https://www.postgresql.org/docs/current/sql-prepare.html).

For example, imagine, that we have an indexed column `is_active` which value is `true` for only 1% of records.
If we write a query `select * from my_table where is_active = true` postgres is smart enough to use `generic` plan which will include index scan as expected.

But what should postgres do if we write a query `select * from my_table where is_active = $1`?
This query should really have 2 different plans for 2 parameter values.
The best postgres can do is to fallback to `custom` plan, which will be evaluated for each query execution.

The developer should be able to inline parameters on which the plan depends and dynamically bind the rest.

**Reply by belovaf** [reply]
For example, SeaORM have `DbBackend::boolean_value()` for this reason.

**Reply by weiznich** [reply]
I'm not sure what you want to add with these comments to the discussion? Are you able to provide actual numbers for this? Otherwise it's just guessing that it might be faster or not, which is really not. 

That said: I'm not totally opposed to add such methods if someone can demonstrate a performance advantage and if someone is willing to put in the work implementing it. You are welcome to contribute here. 

Edit: To be clear here: I personally do not see this as blocking or high priority issue as this can be easily implemented by a third party crate as extension.

**Reply by belovaf** [reply]
Sure.

I have created the simpliest benchmark (for complex queries difference is bigger).

```
create table my_table
(
    id        bigint generated always as identity primary key,
    is_active bool not null
);

insert into my_table(is_active)
select false
from generate_series(1, 1000000) x;

insert into my_table(is_active) values (true);

create index my_table_is_active_idx on my_table(is_active);
```

```
#[bench]
fn select_literal(b: &mut Bencher) {
    let mut con = init_connection();

    b.iter(|| {
        let count = my_table::table
            .filter(my_table::is_active.eq(BoolLiteral(true)))
            .select(count_star())
            .get_result::<i64>(&mut con)
            .unwrap();

        black_box(count);
    });
}

#[bench]
fn select_bind(b: &mut Bencher) {
    let mut con = init_connection();

    b.iter(|| {
        let count = my_table::table
            .filter(my_table::is_active.eq(true))
            .select(count_star())
            .get_result::<i64>(&mut con)
            .unwrap();
        black_box(count);
    });
}
```

Results:
test tests::select_bind    ... bench:     725,810.26 ns/iter (+/- 266,623.54)
test tests::select_literal ... bench:     660,386.51 ns/iter (+/- 148,967.49)

It is 10% difference.

**Reply by belovaf** [reply]
For example, I have tried to repeat the same predicate 15 times for each query, the results are:
test tests::select_bind    ... bench:     795,133.44 ns/iter (+/- 162,428.02)
test tests::select_literal ... bench:     688,388.66 ns/iter (+/- 158,588.53)

And this is already 15% difference.

**Reply by mjanda** [reply]
I will add that if I remember correctly, when you prepare statement with bound boolean variable, PostgreSQL cannot know if it can use partial index over that field. Can you re-run your benchmark with partial index? Thank you

**Reply by belovaf** [reply]
This is not mentioned in postgres documentation. As I see postgres uses the partial index in both cases:
test tests::select_bind    ... bench:     748,491.19 ns/iter (+/- 319,222.16)
test tests::select_literal ... bench:     674,476.44 ns/iter (+/- 158,047.36)

**Reply by weiznich** [reply]
Thanks for providing these numbers. Looks like it might be a meaningful addition to diesel itself. As already written before I do currently not have the time to work on that on my own, but we would appropriate a PR for this.

I think we could go with something like that as function signature:

```rust
fn literal_value<const V: bool, DB: Backend>() -> LiteralValue<V, DB> {}
```

This would allow us to expand that function to also support integers and similar types in the future as soon as we can use generic types for const parameters.

Things I would expect from an implementation: 

* Documentation and tests should be part of the PR
* We need to make sure that `LiteralValue<true>` has a different static query ID than `LiteralValue<false>` as otherwise the prepared statement caching breaks as soon as a user inserts both into the statement cache. (There should be a test for that)
* We need to check if the backend supports the boolean type, therefore this backend parameter in the proposed function signature.

If anyone is willing to work on this we would be happy to receive a PR. If you have problems with your local setup let me know so that I can try to help.
